[AIRFLOW-3958] List tasks upstream work in chain

zhongjiajie opened the pull request:
### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-3958

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

For now on, `airflow.utils.helpers.chain` only support list as downstream, but not upstream. This ticket is to support list as upstream.

Only support

```markdown
task_1 --> task_2 --> task_3 --> task_4

or
                  /--> task_3
task_1 --> task_2 
                  \--> task_4
```

This improvement to be support

```markdown
                  /--> task_3 \
task_1 --> task_2              --> task_5
                  \--> task_4 /
```

### Tests

- [x] My PR adds the following unit tests :

* airflow/tests/utils/helpers/test_chain

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
  - All the public functions and the classes in the PR contain docstrings that explain what it does

### Code Quality

- [x] Passes `flake8`

mik-laj reviewed (approved):
This is useful. This method is not available in the API reference documentation. We need to think about how to add it to  this in future. Now the group of users of this function is very narrow.

Reference: 
https://airflow.readthedocs.io/en/latest/code.html

ashb commented on airflow/utils/helpers.py:
```diff
@@ -154,19 +156,46 @@ def as_flattened_list(iterable):
+                for a, b in zip(task, down_task):
+                    a.set_downstream(b)
+            else:
+                raise AirflowException(
```
https://docs.python.org/2/library/itertools.html#itertools.product is what you want I think?

    # product('ABCD', 'xy') --> Ax Ay Bx By Cx Cy Dx Dy

zhongjiajie replied:
Not really, that the `cross` mean, we already have `cross_downstream` in `airflow.utils.helper`.I am inspired in https://apache-airflow.slack.com/archives/CCPRP7943/p1554633832061000?thread_ts=1554374078.016800&cid=CCPRP7943 , thinking we should have a function to make parallel pipelines work

```
     /  -> t1 -> t3
t0
     \  -> t2 -> t4
```
If we do that, `chain` could do `chain(t0, [t1, t2, t3], [t4, t5, t6])` like

```
     /  -> t1 -> t4
t0      -> t2 -> t5
     \  -> t3 -> t6
```
and could `cross_downstream([t1, t2, t3], [t4, t5, t6])`

```
t1   \     /  -> t4
        \ /
t2    -> X    -> t5
        / \
t3   /      \ -> t6
```

make our dependent more possibilities.

bcb replied:
I agree with @zhongjiajie , it wouldn't make sense to do that.

However I still think you should just zip and join the lists, without checking they're the same size. Some tasks in the bigger list won't be joined to anything - that's fine, the user will realise their mistake.

ashb replied:
Oh yeah, `product` is totally different behaviour.

I'm not so sure that it would be obvious - the "extra" tasks from the long list would instead end up with no deps set. That sounds like it would be more confusing than the behaviour in the PR: an error would show up in the UI.

But I am not a user of this fn so I can't say for certain.

potiuk commented on airflow/utils/helpers.py:
```diff
@@ -150,19 +150,40 @@ def as_flattened_list(iterable):
+            up_task.set_downstream(down_task)
+        elif isinstance(down_task, BaseOperator):
+            down_task.set_upstream(up_task)
+        else:
```
We should check here that these are both iterables and that they have the same length (and throw exception otherwise).

zhongjiajie replied:
We discuss in https://github.com/apache/airflow/pull/4779#issuecomment-480687872 but some of contributors think we should not check list size, WDYT

potiuk replied:
Yeah... Long comments. I think otherwise - the problem is that some operators might be swallowed silently.
This might be misleading.

zhongjiajie replied:
I add throw exception if type not match or length of Iterable not match in `else` section.

potiuk commented on tests/utils/test_helpers.py:
```diff
@@ -248,6 +248,16 @@ def test_cross_downstream(self):
+        self.assertEqual([t4], t2.get_direct_relatives(upstream=False))
+        self.assertEqual([t5], t3.get_direct_relatives(upstream=False))
+        self.assertCountEqual([t4, t5], t6.get_direct_relatives(upstream=True))
+
```
We need more tests showing what happens in case two neighbouring iterables are not the same length.

zhongjiajie replied:
As we discuss above, if we decide add raise Error if list not in the same size, I will add more unit test, thanks

potiuk replied:
I think tests are needed regardless - they should test the behaviour that is description. - but showing possibly different behaviour (swallowing if we decide to do so or error).

zhongjiajie replied:
I add more test on different length of iterable and between not two iterable

potiuk reviewed (changes requested):
I think we also need to document it in the general documentation where we explain how to use upstream and downstream << >>. Around here: https://airflow.readthedocs.io/en/stable/concepts.html#bitshift-composition

feluelle reviewed (approved):
Code and tests LGTM 👍 

+1 on adding docs to this PR. Docs should come always along with a new feature.
@zhongjiajie you want that the feature gets attention - docs would help on that :)

kaxil commented on airflow/utils/helpers.py:
```diff
@@ -150,19 +151,50 @@ def as_flattened_list(iterable):
     """
-    for up_task, down_task in zip(tasks[:-1], tasks[1:]):
-        up_task.set_downstream(down_task)
+    from airflow.models import BaseOperator
```
Let's move the import statement at the top of the file

zhongjiajie replied:
@kaxil I try put import in the top file, but it CI failed. And then Got advice from @BasPH to put `import` inside functions to pass CI

kaxil replied:
Yup, I realised it later. It was to avoid circular imports

kaxil reviewed (commented):
Let's add some docs on it

zhongjiajie committed:
[AIRFLOW-3958] Support list tasks as upstream in chain

helpers.chain only support list as downstream
This PR make list as upstream work, also make
list parallel work, which like below

     / -> t2 -> t4 \
t1 ->               -> t6
     \ -> t3 -> t5 /

zhongjiajie commented on docs/concepts.rst:
```diff
@@ -185,6 +185,9 @@ Bitshift Composition
 *Added in Airflow 1.8*
 
+We recommend you setting operator relationships with bitshift operators rather than ``set_upstream()``
+and ``set_downstream()``.
```
Add this to let users know that we recommend set relationships using bitshift operators

zhongjiajie commented on docs/concepts.rst:
```diff
@@ -248,21 +251,92 @@ Bitshift can also be used with lists. For example:
 .. code:: python
 
-    op1 >> [op2, op3]
+    op1 >> [op2, op3] >> op4
```
Let users know `>>` could work on `list[operatos]` as upstream

zhongjiajie reviewed (commented):
Explain why I change old docs

potiuk reviewed (approved):
LGTM. Looks like quite reasonable way of building more complex chains and it is quite well documented now. @ashb @feluelle @mik-laj @kaxil - what do you think - before we merge it - are you ok with it as well?

ashb reviewed (approved):
LGTM

kaxil commented on airflow/utils/helpers.py:
```diff
@@ -150,19 +151,50 @@ def as_flattened_list(iterable):
 
 def chain(*tasks):
-    """
+    r"""
```
What is this `r` for? Is it a typo or a way to render something?

ashb replied:
> Both string and bytes literals may optionally be prefixed with a letter 'r' or 'R'; such strings are called raw strings and treat backslashes as literal characters

This makes the `\` in the ascii diagram be a backslash. I imagine otherwise it would complain about `\ ` being an invalid escape.

kaxil replied:
Makes sense 👍 Thanks @ashb

zhongjiajie replied:
Thanks ash for the clarification
